backend/src/main/java/ai/giskard/service/ee/LicenseService.java

```diff
@@ -0,0 +1,156 @@
+import org.bouncycastle.crypto.signers.Ed25519Signer;
+import org.bouncycastle.util.encoders.Hex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PostConstruct;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.*;
+
+@Service
+@RequiredArgsConstructor
+public class LicenseService {
+
+    private final Logger log = LoggerFactory.getLogger(LicenseService.class);
+
+    private final FileLocationService fileLocationService;
+
+    private static final String SIGNATURE_KEY = "c947f66224d465b50004c327fc831cff672fc07b540b0613d6f661d0e72d455d";
+
+    @Getter
+    @Setter
+    public class License {
+        private String planName;
+        private String planCode;
+        private List<String> features;
+
+        public Map<FeatureFlagService.FeatureFlag, Boolean> getFeatures() {
+            Map<FeatureFlagService.FeatureFlag, Boolean> map = new HashMap<>();
+
+            for (FeatureFlagService.FeatureFlag featureFlag : FeatureFlagService.FeatureFlag.values()) {
+                map.put(featureFlag, false);
+            }
+
+            if (features != null) {
+                for (String feat : features) {
+                    map.put(FeatureFlagService.FeatureFlag.valueOf(feat), true);
+                }
+            }
+
+            return map;
+        }
+
+        public boolean hasFeature(FeatureFlagService.FeatureFlag flag) {
+            return this.getFeatures().get(flag);
+        }
+    }
+
+    private License license;
+
+    public License getCurrentLicense() {
+        License defaultLicense = new License();
+        defaultLicense.setPlanName("Open Source");
+        return license == null ? defaultLicense : license;
+    }
+
+    /**
+     * Checks if there currently is a license. If yes, parses and loads it.
+     */
+    @PostConstruct
+    public void loadLicense() throws IOException {
+        if (Files.exists(fileLocationService.licensePath())) {
+            String licenseFile = Files.readString(fileLocationService.licensePath());
+            decodeLicense(licenseFile);
+        }
+    }
+
+    /**
+     * Takes a license file content, parses it, validates that it is a proper license.
+     * If it's a real license, saves it in giskard home and updates available feature flags
+     *
+     * @param licenseFile
+     */
+    public void uploadLicense(String licenseFile) throws IOException {
+        decodeLicense(licenseFile);
+        Files.write(fileLocationService.licensePath(), licenseFile.getBytes());
+    }
+
+    private License decodeLicense(String lic) throws IOException {
+        // 1. Remove start/end decorators
+        String encodedPayload = lic.replaceAll("(^-----BEGIN LICENSE FILE-----\\n|\\n|-----END LICENSE FILE-----\\n$)", "");
+
+        // 2. Decode base64 => json
+        byte[] payloadBytes = Base64.getDecoder().decode(encodedPayload);
```

Comment:
I'd extract the different steps into separate methods, it'll be easied to read the main decodeLicense logic on a higher level while being able to dive into submethods if needed